<launch>
  <arg name="run_rviz" default="true" />

  <arg name="INPUT_CLOUD" default="/kinect_head_remote/depth_registered/points" />
  <arg name="sensor_frame" default="head_mount_kinect_rgb_optical_frame" />

  <!-- TODO robot self mask -->


  <!-- hsi color filter -->
  <node name="yellow_calib_hsi_input_relay" pkg="nodelet" type="nodelet"
        args="standalone jsk_topic_tools/Relay">
    <!-- <remap from="~input" to="/extract_indices/output" /> -->
    <remap from="~input" to="$(arg INPUT_CLOUD)" />
  </node>

  <arg name="h_max" value="45" /><!-- 43 -->
  <arg name="h_min" value="30" /><!-- 30 -->
  <arg name="s_max" value="130" /><!-- 101 -->
  <arg name="s_min" value="70" /><!-- 77 -->
  <arg name="i_max" value="240" /><!-- 240 -->
  <arg name="i_min" value="100" /><!-- 156 -->

  <node pkg="nodelet" type="nodelet" name="yellow_calib_hsi_filter"
        args="standalone jsk_pcl/HSIColorFilter" output="screen">
    <!-- <remap from="~input" to="/extract_indices/output" /> -->
    <remap from="~input" to="yellow_calib_hsi_input_relay/output" />
    <!-- <remap from="~input" to="hsi_plane_extraction/output" /> -->
    <remap from="~output" to="yellow_calib_hsi_output" />
    <rosparam>
      use_indices: false
      keep_organized: true
    </rosparam>
    <param name="h_limit_max" value="$(arg h_max)" />
    <param name="h_limit_min" value="$(arg h_min)" />
    <param name="s_limit_max" value="$(arg s_max)" />
    <param name="s_limit_min" value="$(arg s_min)" />
    <param name="i_limit_max" value="$(arg i_max)" />
    <param name="i_limit_min" value="$(arg i_min)" />
  </node>

  <node pkg="nodelet" type="nodelet" name="yellow_calib_hsi_euclidean_clustering"
        args="standalone jsk_pcl/EuclideanClustering" output="screen">
    <remap from="~input" to="yellow_calib_hsi_output" />
    <!-- <remap from="~input" to="hsi_plane_extraction/output" /> -->
    <rosparam>
      tolerance: 0.03
      min_size: 100
      cluster_filter: 1
    </rosparam>
  </node>
  <node name="yellow_calib_hsi_throttle_segmentation" pkg="nodelet" type="nodelet"
        args="standalone jsk_topic_tools/LightweightThrottle"
        output="screen">
    <remap from="~input" to="yellow_calib_hsi_euclidean_clustering/output" />
    <remap from="~output" to="yellow_calib_hsi_euclidean_clustering/output_throttle" />
  </node>
  <node name="yellow_calib_hsi_segmentation_decomposer" pkg="nodelet" type="nodelet"
        args="standalone jsk_pcl/ClusterPointIndicesDecomposer"
        output="screen"
	    clear_params="true">
    <remap from="~input" to="yellow_calib_hsi_output" />
    <!-- <remap from="~input" to="hsi_plane_extraction/output" /> -->
    <!-- <remap from="~target" to="hsi_euclidean_clustering/output" /> -->
    <remap from="~target" to="yellow_calib_hsi_euclidean_clustering/output_throttle" />
    <remap from="~align_planes" to="yellow_calib_hsi_multi_plane_estimate/output_refined_polygon" />
    <remap from="~align_planes_coefficients"
           to="yellow_calib_hsi_multi_plane_estimate/output_refined_coefficients" />
    <rosparam subst_value="true">
      align_boxes: true
      align_boxes_with_plane: false
      target_frame_id: base_link
      use_pca: true
      publish_clouds: false
      publish_tf: true
      sort_by: -cloud_size
    </rosparam>
  </node>

  <group if="$(arg run_rviz)">
    <node pkg="rviz" name="yellow_calib_rviz" type="rviz"
          args="-d $(find jsk_2023_09_cook_from_recipe)/rviz/yellow_calib_test.rviz" />
  </group>

</launch>
